mlir/include/mlir/Dialect/Mesh/IR/MeshOps.td

```diff
@@ -584,11 +584,11 @@ def Mesh_AllReduceOp : Mesh_CollectiveCommunicationOpBase<"all_reduce", [
     ```
   }];
   let arguments = !con(commonArgs, (ins
-    AnyRankedTensor:$input,
+    AnyTypeOf<[AnyMemRef, AnyRankedTensor]>:$input,
```

Comment:
I wonder if this is the way to go, isn't the sharding dialect supposed to be tensor only? I would expect MPI to insert some bufferization dialect ops to get to the memref. Although admittedly, my knowledge here is very flaky.